packages/python/plotly/plotly/express/_core.py

```diff
@@ -1438,42 +1573,32 @@ def build_dataframe(args, constructor):
         value_name = _escape_col_name(columns, "value", [])
         var_name = _escape_col_name(columns, var_name, [])
 
-    if needs_interchanging:
-        try:
-            if wide_mode or not hasattr(args["data_frame"], "select_columns_by_name"):
-                args["data_frame"] = pd.api.interchange.from_dataframe(
-                    args["data_frame"]
-                )
-            else:
-                # Save precious resources by only interchanging columns that are
-                # actually going to be plotted.
-                necessary_columns = {
-                    i for i in args.values() if isinstance(i, str) and i in columns
-                }
-                for field in args:
-                    if args[field] is not None and field in array_attrables:
-                        necessary_columns.update(i for i in args[field] if i in columns)
-                columns = list(necessary_columns)
-                args["data_frame"] = pd.api.interchange.from_dataframe(
-                    args["data_frame"].select_columns_by_name(columns)
-                )
-        except (ImportError, NotImplementedError) as exc:
-            # temporary workaround; developers of third-party libraries themselves
-            # should try a different implementation, if available. For example:
-            # def __dataframe__(self, ...):
-            #   if not some_condition:
-            #     self.to_pandas(...)
-            if hasattr(df_not_pandas, "toPandas"):
-                args["data_frame"] = df_not_pandas.toPandas()
-            elif hasattr(df_not_pandas, "to_pandas_df"):
-                args["data_frame"] = df_not_pandas.to_pandas_df()
-            elif hasattr(df_not_pandas, "to_pandas"):
-                args["data_frame"] = df_not_pandas.to_pandas()
-            else:
-                raise exc
-
-    df_input = args["data_frame"]
+    if isinstance(args["data_frame"], nw.DataFrame) and needs_interchanging:
+        # Interchange to PyArrow
```

Comment:
I can't quite grok why this PyArrow interchange step is necessary if PyArrow is supported by Narwhals -- can you explain?

Reply:
This is for interchange-level inputs (duckdb, ibis, anything that implemented the dataframe interchange protocol) that we don't have full API support for. So, once we've selected the necessary columns, we convert to pyarrow and then continue from there